Refactor parts of Irrlicht's material system

grorp opened the pull request:
This is the Minetest side of minetest/irrlicht#204 ("Refactor parts of the material system and set min filter to bilinear for 2D graphics"). The actual change is in the Irrlicht PR, this is "just" refactoring I did for that change.

Together with minetest/irrlicht#204, this PR fixes #13621. Because it exposes all OpenGL texture filtering options to Minetest, it also makes #13108 trivial to fix.

See minetest/irrlicht#204 for more information.

## To do

This PR is a Ready for Review.

## How to test

Use Minetest on all platforms and with all Irrlicht video drivers and verify that everything looks the same as before, with one exception: Images rendered with a lower resolution than their original resolution in GUIs look better. The difference is very visible with #13606, for example.

grorp commented:
I see a problem here. If the Irrlicht PR is merged first, all Minetest builds will fail until this one is merged too. If this PR is merged first, all Minetest builds will fail until the Irrlicht PR is merged too.

As the changes affect many lines of code, I don't think using #defines in Minetest to make Minetest compatible with both Irrlicht versions is possible. The only solution is see is to merge both PRs and bump the IrrlichtMt tag at once. Does anyone have a better idea? Is this even possible?

sfan5 commented:
> If the Irrlicht PR is merged first, all Minetest builds will fail until this one is merged too.

They won't as the CI builds use a fixed tag from IrrlichtMt. This will only be an issue for users cloning both repos from the latest state.
this is also why the following suggestion exists: #13215

> The only solution is see is to merge both PRs and bump the IrrlichtMt tag at once.

Given the current situation this is still the best option.

grorp commented:
Just to be clear, all this refactoring wasn't strictly required for separating the min and mag filters and setting the min filter to bilinear for 2D graphics.

I could have done this with a lot less changes, but that would have resulted in an inconsistent material API. I made all these changes because I wanted a consistent API for setting material properties.

Before this PR:
- `SMaterialLayer` has two boolean fields, `BilinearFilter` and `TrilinearFilter`, which select both the min and the mag filter
- you set them via `SMaterial::setFlag`, which automatically sets them on all texture layers for you

After this PR:
- `SMaterialLayer` has two enum fields, `MinFilter` (nearest, bilinear or trilinear) and `MagFilter` (nearest or bilinear)
- ❌ setting them via `SMaterial::setFlag`? impossible because they are no booleans
- ❌ manually setting them on all texture layers? would cause lots of code duplication
- ✅ you set them in a lambda passed to the new `SMaterial::forEachTexture` function

However, setting a per-texture material flag on all texture layers was the only real use of `SMaterial::setFlag`. All other uses could trivially be replaced by something like `material.<name of the flag> = <value>`, i.e. `setFlag` was a simple setter function. Therefore, I replaced all of its uses and removed it.

Desour reviewed (approved):
Code looks fine. LGTM.

sfan5 commented on src/client/clientmap.cpp:
```diff
@@ -842,32 +842,29 @@ void ClientMap::renderMap(video::IVideoDriver* driver, s32 pass)
-				m_cache_anistropic_filter);
-			material.setFlag(video::EMF_WIREFRAME,
-				m_control.show_wireframe);
+			material.forEachTexture([this] (video::SMaterialLayer &tex) {
```
consider using `auto` here

grorp replied:
Now using `auto` where applicable.

sfan5 reviewed (approved):
tested together with #13606, works

grorp commented:
Rebased, removed "(Minetest side)" from the commit descriptions, squashed a bit.
